File: Repositorio/engenharia-de-software-com-ia-aplicada-main/modulo09-processamento-de-dados-e-fine-tuning-de-modelos/modulo-04-lora-e-peft/rank-adapter-comparison-tool.js
/**
 * Ahirton Lopes · Fine-Tuning Toolkit
 * Artefato de Demo - Módulo 4.3 (Demo, parte 2 -- companion)
 *
 * Roda o exemplo difícil do TP (Boa Vista Reparos Automotivos, com dois
 * valores distratores antes do valor certo) sem adaptador e com os três
 * postos testados no módulo -- rank 4, 8 (reaproveita o adaptador real do
 * Módulo 4.2) e 16. Mesmo modelo, mesmo prompt, saída real dos quatro.
 *
 * Uso: node rank-adapter-comparison-tool.js
 *
 * Par oficial desta disciplina: rank-adapter-comparison-tool.js (oficial) /
 * rank_adapter_comparison_tool.py (referência espelhada).
 */

'use strict';

const assert = require('assert').strict;
const path = require('path');
const { spawnSync } = require('child_process');

const MODELO_BASE = 'mlx-community/gemma-4-e2b-it-bf16';
const MAX_TOKENS = 80;

const ADAPTERS = {
  'rank 4': path.join(__dirname, 'mlx-adapters-rank4'),
  'rank 8': path.join(__dirname, 'mlx-adapters'),
  'rank 16': path.join(__dirname, 'mlx-adapters-rank16'),
};

const PROMPT = `Extraia segurado, placa e valor do orçamento de oficina abaixo.

BOA VISTA REPAROS AUTOMOTIVOS CNPJ 21.098.765/0001-32 Rua dos Mecanicos 310 Segurado: Ricardo Alves Monteiro Placa do veiculo: JBR-9021 Data do sinistro: 09/06/2026 Valor das pecas: R$ 1.850,00 Valor da mao de obra: R$ 970,00 Valor total do orcamento: R$ 2.820,00`;

const GABARITO = { segurado: 'Ricardo Alves Monteiro', placa: 'JBR-9021', valor: 2820 };

/* ===== montar argumentos e rodar ===== */

function montarArgumentos(adapterPath) {
  const args = ['-m', 'mlx_lm', 'generate', '--model', MODELO_BASE, '--prompt', PROMPT, '--max-tokens', String(MAX_TOKENS)];
  if (adapterPath) args.push('--adapter-path', adapterPath);
  return args;
}

function parsearSaida(textoSaida) {
  const blocos = textoSaida.split('==========');
  if (blocos.length < 3) throw new Error('saída fora do formato esperado');
  const texto = blocos[1].trim();
  const tokens = Number((textoSaida.match(/Generation: (\d+) tokens/) || [])[1] || 0);
  let json = null;
  try { json = JSON.parse(texto); } catch { /* não é JSON, tudo bem */ }
  return { texto, tokens, json };
}

function baterComGabarito(json) {
  return !!json && Object.keys(GABARITO).every((k) => json[k] === GABARITO[k]);
}

function rodarReal(adapterPath, spawnFn = spawnSync) {
  const r = spawnFn('python3', montarArgumentos(adapterPath), { encoding: 'utf-8', maxBuffer: 10 * 1024 * 1024 });
  if (r.error) throw r.error;
  if (r.status !== 0) throw new Error(`mlx_lm generate falhou (${r.status}):\n${r.stderr}`);
  return `${r.stdout}\n${r.stderr}`;
}

/* ===== testes -- contra saída real capturada em 2026-09-04 ===== */

const SAIDA_SEM_ADAPTADOR = `==========
<|channel>thought
Here's a thinking process to extract the requested information:

1.  **Analyze the Request:** The user wants to extract three specific pieces of information from the provided text (an auto repair quote/budget):
==========
Prompt: 155 tokens, 211.031 tokens-per-sec
Generation: 80 tokens, 50.460 tokens-per-sec
Peak memory: 9.447 GB`;

const SAIDA_RANK4 = `==========
{"segurado":"Ricardo Alves Monteiro","placa":"JBR-9021","valor":2820}
==========
Prompt: 155 tokens, 218.733 tokens-per-sec
Generation: 27 tokens, 41.540 tokens-per-sec
Peak memory: 9.447 GB`;

let total = 0;
let falhas = 0;
function testar(desc, fn) {
  total += 1;
  try { fn(); console.log(`  [OK] ${desc}`); }
  catch (e) { falhas += 1; console.log(`  [FALHOU] ${desc}\n           ${e.message}`); }
}

function rodarTestes() {
  console.log('== Testes ==');
  testar('sem adaptador: bate no limite de tokens, não chega a JSON', () => {
    const r = parsearSaida(SAIDA_SEM_ADAPTADOR);
    assert.equal(r.tokens, 80);
    assert.equal(r.json, null);
  });
  testar('rank 4: JSON exato batendo com o gabarito', () => {
    const r = parsearSaida(SAIDA_RANK4);
    assert.deepEqual(r.json, GABARITO);
  });
  testar('baterComGabarito confirma e rejeita corretamente', () => {
    assert.equal(baterComGabarito(GABARITO), true);
    assert.equal(baterComGabarito({ segurado: 'outro' }), false);
  });
  console.log(`\n${total - falhas}/${total} testes passaram.`);
  if (falhas > 0) process.exitCode = 1;
}

/* ===== comparação real ===== */

function rodarComparacaoReal() {
  console.log('\n===== Boa Vista Reparos Automotivos: sem adaptador vs. rank 4/8/16 =====\n');

  const resultados = {};
  const falhas = [];

  console.log('--- sem adaptador ---');
  try {
    const semAdaptador = parsearSaida(rodarReal(null));
    console.log(`${semAdaptador.tokens} tokens, bate com o gabarito? ${baterComGabarito(semAdaptador.json)}`);
    resultados['sem adaptador'] = semAdaptador;
  } catch (erro) {
    console.log(`[FALHOU] ${erro.message}`);
    falhas.push({ nome: 'sem adaptador', erro: erro.message });
  }

  for (const [nome, caminho] of Object.entries(ADAPTERS)) {
    console.log(`--- ${nome} ---`);
    try {
      const r = parsearSaida(rodarReal(caminho));
      console.log(`${r.tokens} tokens, bate com o gabarito? ${baterComGabarito(r.json)}`);
      if (r.json) console.log(JSON.stringify(r.json));
      resultados[nome] = r;
    } catch (erro) {
      console.log(`[FALHOU] ${erro.message}`);
      falhas.push({ nome, erro: erro.message });
    }
  }

  if (falhas.length > 0) {
    console.log(`\n${falhas.length}/${Object.keys(ADAPTERS).length + 1} configuração(ões) falharam ao rodar mlx_lm generate.`);
  }
  if (Object.keys(resultados).length === 0) {
    throw new Error('Nenhuma configuração rodou com sucesso -- verifique o modelo/adaptadores locais.');
  }
  return resultados;
}

if (require.main === module) {
  rodarTestes();
  try {
    rodarComparacaoReal();
  } catch (erro) {
    console.error('Erro:', erro.message);
    process.exitCode = 1;
  }
}

module.exports = { ADAPTERS, PROMPT, GABARITO, montarArgumentos, parsearSaida, baterComGabarito, rodarReal, rodarComparacaoReal };

/*
 * Ahirton Lopes · Fine-Tuning Toolkit - UNIPDS: Processamento de Dados e Fine-Tuning de Modelos
 * Prof. Ahirton Lopes, Ph.D. - GDE AI, Microsoft MVP, Senior Manager
 */

File: Repositorio/engenharia-de-software-com-ia-aplicada-main/modulo09-processamento-de-dados-e-fine-tuning-de-modelos/modulo-04-lora-e-peft/rank_adapter_comparison_tool.py
"""
Ahirton Lopes - Fine-Tuning Toolkit
Artefato de Demo - Módulo 4.3 (Demo, parte 2 -- companion, referência espelhada)

Roda o exemplo difícil do TP (Boa Vista Reparos Automotivos, com dois
valores distratores antes do valor certo) sem adaptador e com os três
postos testados no módulo -- rank 4, 8 (reaproveita o adaptador real do
Módulo 4.2) e 16. Mesmo modelo, mesmo prompt, saída real dos quatro.

Uso: python3 rank_adapter_comparison_tool.py
"""

import json
import os
import subprocess

MODELO_BASE = "mlx-community/gemma-4-e2b-it-bf16"
MAX_TOKENS = 80

_AQUI = os.path.dirname(os.path.abspath(__file__))
ADAPTERS = {
    "rank 4": os.path.join(_AQUI, "mlx-adapters-rank4"),
    "rank 8": os.path.join(_AQUI, "mlx-adapters"),
    "rank 16": os.path.join(_AQUI, "mlx-adapters-rank16"),
}

PROMPT = (
    "Extraia segurado, placa e valor do orçamento de oficina abaixo.\n\n"
    "BOA VISTA REPAROS AUTOMOTIVOS CNPJ 21.098.765/0001-32 Rua dos Mecanicos 310 "
    "Segurado: Ricardo Alves Monteiro Placa do veiculo: JBR-9021 Data do sinistro: "
    "09/06/2026 Valor das pecas: R$ 1.850,00 Valor da mao de obra: R$ 970,00 "
    "Valor total do orcamento: R$ 2.820,00"
)

GABARITO = {"segurado": "Ricardo Alves Monteiro", "placa": "JBR-9021", "valor": 2820}


def montar_argumentos(adapter_path=None):
    args = ["-m", "mlx_lm", "generate", "--model", MODELO_BASE, "--prompt", PROMPT, "--max-tokens", str(MAX_TOKENS)]
    if adapter_path:
        args += ["--adapter-path", adapter_path]
    return args


def parsear_saida(texto_saida):
    blocos = texto_saida.split("==========")
    if len(blocos) < 3:
        raise ValueError("saída fora do formato esperado")
    texto = blocos[1].strip()
    import re
    m = re.search(r"Generation: (\d+) tokens", texto_saida)
    tokens = int(m.group(1)) if m else 0
    try:
        saida_json = json.loads(texto)
    except (json.JSONDecodeError, ValueError):
        saida_json = None
    return {"texto": texto, "tokens": tokens, "json": saida_json}


def bater_com_gabarito(saida_json):
    return bool(saida_json) and all(saida_json.get(k) == v for k, v in GABARITO.items())


def rodar_real(adapter_path=None, runner=subprocess.run):
    r = runner(["python3"] + montar_argumentos(adapter_path), capture_output=True, text=True)
    if r.returncode != 0:
        raise RuntimeError(f"mlx_lm generate falhou ({r.returncode}):\n{r.stderr}")
    return f"{r.stdout}\n{r.stderr}"


SAIDA_SEM_ADAPTADOR = """==========
<|channel>thought
Here's a thinking process to extract the requested information:

1.  **Analyze the Request:** The user wants to extract three specific pieces of information from the provided text (an auto repair quote/budget):
==========
Prompt: 155 tokens, 211.031 tokens-per-sec
Generation: 80 tokens, 50.460 tokens-per-sec
Peak memory: 9.447 GB"""

SAIDA_RANK4 = """==========
{"segurado":"Ricardo Alves Monteiro","placa":"JBR-9021","valor":2820}
==========
Prompt: 155 tokens, 218.733 tokens-per-sec
Generation: 27 tokens, 41.540 tokens-per-sec
Peak memory: 9.447 GB"""

_total = 0
_falhas = 0


def testar(desc, fn):
    global _total, _falhas
    _total += 1
    try:
        fn()
        print(f"  [OK] {desc}")
    except Exception as e:  # noqa: BLE001
        _falhas += 1
        print(f"  [FALHOU] {desc}\n           {e}")


def rodar_testes():
    print("== Testes ==")

    def t1():
        r = parsear_saida(SAIDA_SEM_ADAPTADOR)
        assert r["tokens"] == 80
        assert r["json"] is None
    testar("sem adaptador: bate no limite de tokens, não chega a JSON", t1)

    def t2():
        r = parsear_saida(SAIDA_RANK4)
        assert r["json"] == GABARITO
    testar("rank 4: JSON exato batendo com o gabarito", t2)

    def t3():
        assert bater_com_gabarito(GABARITO) is True
        assert bater_com_gabarito({"segurado": "outro"}) is False
    testar("bater_com_gabarito confirma e rejeita corretamente", t3)

    print(f"\n{_total - _falhas}/{_total} testes passaram.")
    return _falhas == 0


def rodar_comparacao_real():
    print("\n===== Boa Vista Reparos Automotivos: sem adaptador vs. rank 4/8/16 =====\n")

    resultados = {}
    falhas = []

    print("--- sem adaptador ---")
    try:
        sem_adaptador = parsear_saida(rodar_real(None))
        print(f"{sem_adaptador['tokens']} tokens, bate com o gabarito? {bater_com_gabarito(sem_adaptador['json'])}")
        resultados["sem adaptador"] = sem_adaptador
    except Exception as erro:
        print(f"[FALHOU] {erro}")
        falhas.append({"nome": "sem adaptador", "erro": str(erro)})

    for nome, caminho in ADAPTERS.items():
        print(f"--- {nome} ---")
        try:
            r = parsear_saida(rodar_real(caminho))
            print(f"{r['tokens']} tokens, bate com o gabarito? {bater_com_gabarito(r['json'])}")
            if r["json"]:
                print(json.dumps(r["json"]))
            resultados[nome] = r
        except Exception as erro:
            print(f"[FALHOU] {erro}")
            falhas.append({"nome": nome, "erro": str(erro)})

    if falhas:
        print(f"\n{len(falhas)}/{len(ADAPTERS) + 1} configuracao(oes) falharam ao rodar mlx_lm generate.")
    if not resultados:
        raise RuntimeError("Nenhuma configuracao rodou com sucesso -- verifique o modelo/adaptadores locais.")
    return resultados


if __name__ == "__main__":
    ok = rodar_testes()
    try:
        rodar_comparacao_real()
    except RuntimeError as erro:
        print(f"Erro: {erro}")
        ok = False
    if not ok:
        raise SystemExit(1)

# Ahirton Lopes - Fine-Tuning Toolkit - UNIPDS: Processamento de Dados e Fine-Tuning de Modelos
# Prof. Ahirton Lopes, Ph.D. - GDE AI, Microsoft MVP, Senior Manager
